pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,821 @@
+	} else if ps.Spec.Packet.TransportHeader.ICMP != nil {
+		isICMP = true
+	}
+
+	if packet.IPProto == 0 || packet.IPProto == protocol.Type_ICMP || packet.IPProto == protocol.Type_IPv6ICMP {
+		isICMP = true
+	}
+
+	if isICMP {
+		if packet.IsIPv6 {
+			packet.IPProto = protocol.Type_IPv6ICMP
+
+		} else {
+			packet.IPProto = protocol.Type_ICMP
+		}
+	}
+
+	return packet, nil
+}
+
+func (c *Controller) syncPacketSampling(psName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing PacketSampling for %s. (%v)", psName, time.Since(startTime))
+	}()
+
+	ps, err := c.packetSamplingLister.Get(psName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketSampling(psName)
+			return nil
+		}
+		return err
+	}
+
+	switch ps.Status.Phase {
+	case crdv1alpha1.PacketSamplingRunning:
+		if ps.Status.DataplaneTag != 0 {
+			start := false
+			c.runningPacketSamplingsMutex.Lock()
+			if _, ok := c.runningPacketSamplings[ps.Status.DataplaneTag]; !ok {
+				start = true
+			}
+			c.runningPacketSamplingsMutex.Unlock()
+			if start {
+				err = c.startPacketSampling(ps)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for packet %s", ps.Status.DataplaneTag, ps.Name)
+		}
+	default:
+		c.cleanupPacketSampling(psName)
+	}
+	return err
+
+}
+
+type uploader interface {
+	upload(addr string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error
+}
+
+type sftpUploader struct {
+}
+
+func (uploader *sftpUploader) upload(address string, path string, config *ssh.ClientConfig, packetFile io.Reader) error {
+	conn, err := ssh.Dial("tcp", address, config)
+	if err != nil {
+		return fmt.Errorf("error when connecting to fs server: %w", err)
+	}
+	sftpClient, err := sftp.NewClient(conn)
+	if err != nil {
+		return fmt.Errorf("error when setting up sftp client: %w", err)
+	}
+	defer func() {
+		if err := sftpClient.Close(); err != nil {
+			klog.ErrorS(err, "Error when closing sftp client")
+		}
+	}()
+	targetFile, err := sftpClient.Create(path)
+	if err != nil {
+		return fmt.Errorf("error when creating target file on remote: %v", err)
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			klog.ErrorS(err, "Error when closing target file on remote")
+		}
+	}()
+	if written, err := io.Copy(targetFile, packetFile); err != nil {
+		return fmt.Errorf("error when copying target file: %v, written: %d", err, written)
+	}
+	klog.InfoS("Successfully upload file to path", "filePath", path)
+	return nil
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func parseUploadUrl(uploadUrl string) (*url.URL, error) {
+	parsedURL, err := url.Parse(uploadUrl)
+	if err != nil {
+		parsedURL, err = url.Parse("sftp://" + uploadUrl)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if parsedURL.Scheme != "sftp" {
```

Comment:
According to your comment "fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path",  this seems incorrect if the url is "10.92.23.154:22/path", the scheme will be empty?

This can also be a shared function with support bundle controller. Please reorganize them.

Reply:
in the case of `10.92.23.154:22/path` , the function will try again with a `sftp` prefix.